src/secret.rs

```diff
@@ -0,0 +1,184 @@
+/// This type has a safe `Debug` implementation that does not leak information.
+/// Additionally, it implements `Drop` to zeroize the memory to make leaking
+/// passwords less likely.
+#[derive(Clone)]
+pub struct PrivateKey(Zeroizing<ZeroizeSecretKey>);
+
+impl PrivateKey {
+    /// Creates a new private key from raw bytes.
+    pub fn from_raw(raw: [u8; 32]) -> Result<Self, InvalidPrivateKey> {
+        PrivateKey::from_slice(&raw)
+    }
+
+    /// Creates a new private key from a slice of bytes.
+    pub fn from_slice<B: AsRef<[u8]>>(raw: B) -> Result<Self, InvalidPrivateKey> {
+        let secret_key = SecretKey::from_slice(raw.as_ref())?;
+        Ok(PrivateKey(Zeroizing::new(secret_key.into())))
+    }
+
+    /// Creates a new private key from a hex string representation. Accepts hex
+    /// string with or without leading `"0x"`.
+    pub fn from_hex_str<S: AsRef<str>>(s: S) -> Result<Self, InvalidPrivateKey> {
+        let hex_str = {
+            let s = s.as_ref();
+            if s.starts_with("0x") {
+                &s[2..]
+            } else {
+                s
+            }
+        };
+        let secret_key = SecretKey::from_str(hex_str)?;
+        Ok(PrivateKey(Zeroizing::new(secret_key.into())))
+    }
+
+    /// Gets the public address for a given private key.
+    pub fn public_address(&self) -> Address {
+        let secp = Secp256k1::signing_only();
+        let public_key = PublicKey::from_secret_key(&secp, &*self).serialize_uncompressed();
+
+        // NOTE: An ethereum address is the last 20 bytes of the keccak hash of
+        //   the public key. Note that `libsecp256k1` public key is serialized
+        //   into 65 bytes as the first byte is always 0x04 as a tag to mark a
+        //   uncompressed public key. Discard it for the public address
+        //   calculation.
+        debug_assert_eq!(public_key[0], 0x04);
+        let hash = hash::keccak256(&public_key[1..]);
+
+        Address::from_slice(&hash[12..])
+    }
+}
+
+impl FromStr for PrivateKey {
+    type Err = InvalidPrivateKey;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        PrivateKey::from_hex_str(s)
+    }
+}
+
+impl Deref for PrivateKey {
+    type Target = SecretKey;
+
+    fn deref(&self) -> &Self::Target {
+        &(self.0).0
+    }
+}
+
+impl Debug for PrivateKey {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_tuple("PrivateKey")
+            .field(&self.public_address())
+            .finish()
+    }
+}
+
+/// An internal type that allows us to implement `Zeroize` on `SecretKey`. This
+/// allows `PrivateKey` to correctly zeroize (almost, we use the `ONE_KEY`
+/// instead of `0`s since it is the first valid key) in a way that does not get
+/// optimized away by the compiler or get access reordered.
+///
+/// For more information, consult the `zeroize` crate
+/// [`README`](https://github.com/iqlusioninc/crates/tree/develop/zeroize).
+#[derive(Clone, Copy)]
+struct ZeroizeSecretKey(SecretKey);
```

Comment:
I'm a bit confused that we need a double Zeroize (`Zeroizing<ZeorizeSecretKey>`). Maybe `ZeroizeableSecretKey` is a better name as by itself it seems to not be zeroizing?

Reply:
`Zeroize` is the trait, `Zeroizing` is a wrapper struct that makes sure to call `Zeroize::zeroize` on `Drop`. The naming can be a bit confusing :confused: